### PR TITLE
Fixes #17427 - Enable OSTree Upstream Sync Depth

### PR DESCRIPTION
**Pull request by parthaa**
This commit enables one to set the Upstream Sync Policy and Sync depth.
Basically with this change one is now able to say "Download complete version history
when syncing an ostree repo"

### CODE REVIEW

**Comment by parthaa** [comment]
[test]

**Comment by parthaa** [comment]
Blocked by https://github.com/Katello/runcible/pull/188

**Comment by parthaa** [comment]
**Background**
Katello today when syncing an upstream ostree feed, syncs only the latest versions (aka branch heads). This means if a user for example synced RH Atomic OStree repo they are only going to the get the latest version of RHEL Atomic Host. Many users have wanted the ability to revert to prior versions of  RHEL Atomic Host tree. OSTree + Pulp recently have added ability to set a "depth" value on both the importer and distributor configurations in ostree to remedy this. One can configure the "traversal depth" in pulp to "pull all versions", "pull the last 5 versions", "pull the latest version" of the upstream feed.

**Simple way to see this in bash**
- yum install ostree
```bash
$ mkdir /tmp/repo
$ cd /tmp/repo
$ ostree init --repo=.
# point it to the ostree repo in https://partha.fedorapeople.org/test-repos/ostree-zoo/
$ ostree --repo=. remote add --set=--set=tls-permissive=true --set=gpg-verify=false all https://partha.fedorapeople.org/test-repos/ostree-zoo/ fedora-atomic/f21/x86_64/updates-testing/docker-host

$ ostree pull all --repo=. --depth=0
255 metadata, 660 content objects fetched; 228 KiB transferred in 4 seconds                                                                                                                                        
$ ostree log all:fedora-atomic/f21/x86_64/updates-testing/docker-host
commit b954defc0c21b23705bf5776a78ca434b2569260e64e19fb61bdbd5526b3bf0f
Date:  2016-03-03 15:42:30 +0000
Version: 23.3

    testing

<< History beyond this commit not fetched >>

# next pull with depth of 1
$ ostree pull all --repo=. --depth=1
115 metadata, 3 content objects fetched; 22 KiB transferred in 2 seconds

$ ostree log all:fedora-atomic/f21/x86_64/updates-testing/docker-host
commit b954defc0c21b23705bf5776a78ca434b2569260e64e19fb61bdbd5526b3bf0f
Date:  2016-03-03 15:42:30 +0000
Version: 23.3

    testing

commit b5478067c6ed79acde91d09cd3effcd12885a3cc749c390735de18a91ee75c9c
Date:  2016-03-03 15:42:30 +0000
Version: 23.2

    testing

# next pull with depth of -1 or all history 
$ ostree pull all --repo=. --depth=1
20 metadata, 2 content objects fetched; 2 KiB transferred in 2 seconds   

$ ostree log all:fedora-atomic/f21/x86_64/updates-testing/docker-host
commit b954defc0c21b23705bf5776a78ca434b2569260e64e19fb61bdbd5526b3bf0f
Date:  2016-03-03 15:42:30 +0000
Version: 23.3

    testing

commit b5478067c6ed79acde91d09cd3effcd12885a3cc749c390735de18a91ee75c9c
Date:  2016-03-03 15:42:30 +0000
Version: 23.2

    testing

commit 0dfe48948b5ecaf74f1864b4cb43a991ac7ab7e4ca5bbab36ac30b6596d81799
Date:  2016-03-03 15:42:30 +0000
Version: 23.1

    testing
```

The above shell snippet shows how the more commits/versions  got added as we changed the depth value. 
We need to facilitate this in Katello both in UI and API.


**Intent**
This PR enables users to set an "OSTree upstream sync policy" where the user selects one of "Latest only", vs "All History", "Custom Depth" specifying a custom value for a repository.

**On the Model Side**

- Addition of "ostree_upstream_sync_policy" and "ostree_upstream_sync_depth" to Repository Object. This will affect the syncing from upstream. The valid values for the ostree_upstream_sync_policies are "latest", "all", "custom" and custom will require one to set a value for ostree_upstream_sync_depth

- Forcing an "ostree_publish_depth" of -1 so that pulp publishes everything it syncs and hence it being available to end user
- Forcing an "ostree_capsule_sync_depth" of -1 so that the 

**API**

- Additional parameters in the create/update of repositories_controller to set _ostree_upstream_sync_policy_ and _ostree_upstream_sync_depth_


**UI**

- New options to set an "Ostree Upstream Sync Policy" create an ostree repo

01-create-repo

- Details screen
02-edit-repo

- Edit repos
03-edit-repo

- List repos
04-list-repos

Note that the UI metrics of a sync operation are going to look identical whether you are syncing using All History/Latest Only/Or Custom even though these mean different traversal depth. This is because pulp does not capture "all" the results of the sync in its mongo db. It only captures the "branch heads" and makes units out of them. So Katello has no way of knowing whether all versions got synced. This is being addressed in https://pulp.plan.io/issues/2594


**How To Test This PR**

There are multiple ways to test this pr.
- Try selecting a depth other than "Latest" like "All History"/"Custom" (with a value) and try looking at the repo created in pulp admin. You should see something that looks like

```bash
$ pulp ostree repo list --detail --repo-id=09ac74e0-3a9d-4e24-b441-b0df2fbafe56
Id:                   09ac74e0-3a9d-4e24-b441-b0df2fbafe56
.....
....
Importers:            
  Config:               
    Depth:          -1 # If you selected all history
    Feed: ....
......
Distributors:         
  Auto Publish:         True
  Config:               
    Depth:         -1 # Always -1
.........
```
- Try setting the upstream sync depth and policy via hammer and make it looks ok


**Want to verify that the sync worked**
Now that you have a repo that you created with one of the sync policies and ran the sync operation you might want to examine the contents of the repo.

This is slightly tricky because OSTree content is always protected even for custom repos. Hence to see the contents of the repo you need to provision an atomic host and have that pointed to katello and see in the client. That being said there is a hacky but easier option

Hacky But Easier option
- Disable repo authorization by setting ```enabled: false``` in /etc/pulp/repo_auth.conf
- Optionalize SSL Verification by setting ```SSLVerifyClient optional``` in /etc/httpd/conf.d/pulp_ostree.conf
- Restart Httpd
- Copy the "Published At" url in repo details page of your repo.
05-url-repo
- Finally you are ready to run the actual verification steps

```bash
$ mkdir /tmp/repo1
$ cd /tmp/repo1
$ ostree init --repo=.
# point it to the ostree repo in Publised At -  in my case
# https://obelix.example.com/pulp/ostree/web/Default_Organization/Library/custom/baked/all 
# Its a repo with 'all history' synced
#
$ ostree --repo=. remote add --set=--set=tls-permissive=true --set=gpg-verify=false all https://obelix.example.com/pulp/ostree/web/Default_Organization/Library/custom/baked/all  fedora-atomic/f21/x86_64/updates-testing/docker-host

$ ostree pull all --repo=. --depth=-1 # make sure depth =-1 , you want to capture every thing pulp synced
....
# Finally check the logs

$ ostree log all:fedora-atomic/f21/x86_64/updates-testing/docker-host
commit b954defc0c21b23705bf5776a78ca434b2569260e64e19fb61bdbd5526b3bf0f
Date:  2016-03-03 15:42:30 +0000
Version: 23.3

    testing

commit b5478067c6ed79acde91d09cd3effcd12885a3cc749c390735de18a91ee75c9c
Date:  2016-03-03 15:42:30 +0000
Version: 23.2

    testing

commit 0dfe48948b5ecaf74f1864b4cb43a991ac7ab7e4ca5bbab36ac30b6596d81799
Date:  2016-03-03 15:42:30 +0000
Version: 23.1

    testing
```

Depending on whether you synced as "Latest", "Custom", "All" you should see different answers.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/details/repository-details-info.controller.js [hunk 1]
```diff
@@ -100,10 +101,28 @@ angular.module('Bastion.repositories').controller('RepositoryDetailsInfoControll
...
+
+angular.module('Bastion.components.formatters').filter('ostreeUpstreamSyncPolicyFilter', ['translate', 'OstreeUpstreamSyncPolicy', function (translate, OstreeUpstreamSyncPolicy) {
+    return function (displayValue, repository) {
+      var policy = repository["ostree_upstream_sync_policy"];
+      if ( policy === "custom") {
+          return OstreeUpstreamSyncPolicy.syncPolicyName(policy, repository["ostree_upstream_sync_depth"]);
+      }
+      return OstreeUpstreamSyncPolicy.syncPolicyName(policy);
+    };
+}]);
```
Comment (jlsherrill) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/details/repository-details-info.controller.js]: shouldn't this be in its own file?

**Comment by jlsherrill** [comment]
runcible-1.10.0 has been released, make sure to update the gem requires for that

**Comment by parthaa** [comment]
> runcible-1.10.0 has been released, make sure to update the gem requires for that

@jlsherrill updated.

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 2]
```diff
@@ -151,6 +153,8 @@ def create
       repository.verify_ssl_on_sync = ::Foreman::Cast.to_bool(repo_params[:verify_ssl_on_sync]) if repo_params.key?(:verify_ssl_on_sync)
       repository.upstream_username = repo_params[:upstream_username] if repo_params.key?(:upstream_username)
       repository.upstream_password = repo_params[:upstream_password] if repo_params.key?(:upstream_password)
+      repository.ostree_upstream_sync_policy = repo_params[:ostree_upstream_sync_policy] if repo_params.key?(:ostree_upstream_sync_policy)
+      repository.ostree_upstream_sync_depth = repo_params[:ostree_upstream_sync_depth] if repo_params.key?(:ostree_upstream_sync_depth)
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb]: nitpick  You could use   repo_params.fetch(:ostree_upstream_sync_policy, nil)    here for these two lines

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 3]
```diff
@@ -411,6 +417,7 @@ def find_gpg_key
 
     def repository_params
       keys = [:download_policy, :mirror_on_sync, :verify_ssl_on_sync, :upstream_password, :upstream_username]
+      keys += [:ostree_upstream_sync_depth, :ostree_upstream_sync_policy]
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb]: why not just put this in the previous line's array?

Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb]: unless you meant to add an if repo.ostree?  on the end

**Pull request merged**